File: src/main/java/me/miki/shindo/utils/IOUtils.java
package me.miki.shindo.utils;

import me.miki.shindo.logger.ShindoLogger;
import me.miki.shindo.utils.transferable.ImageTransferable;
import net.minecraft.client.Minecraft;
import net.minecraft.util.ResourceLocation;

import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.StringSelection;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Files;

public class IOUtils {

    private static final Minecraft mc = Minecraft.getMinecraft();

    public static void copyStringToClipboard(String s) {
        StringSelection stringSelection = new StringSelection(s);
        getToolkit().getSystemClipboard().setContents(stringSelection, null);
    }

    public static String getStringFromClipboard() {
        try {
            return getToolkit().getSystemClipboard().getContents(null).getTransferData(DataFlavor.stringFlavor).toString();
        } catch (Exception e) {
            return null;
        }
    }

    public static void copyImageToClipboard(Image image) {
        ImageTransferable imageSelection = new ImageTransferable(image);
        getToolkit().getSystemClipboard().setContents(imageSelection, null);
    }

    public static Image getImageFromClipboard() {
        try {
            return (Image) getToolkit().getSystemClipboard().getContents(null).getTransferData(DataFlavor.imageFlavor);
        } catch (Exception e) {
            return null;
        }
    }

    private static Toolkit getToolkit() {
        return Toolkit.getDefaultToolkit();
    }

    public static ByteBuffer resourceToByteBuffer(ResourceLocation location) {

        try {
            byte[] bytes = org.apache.commons.io.IOUtils.toByteArray(mc.getResourceManager().getResource(location).getInputStream());

            ByteBuffer data = ByteBuffer.allocateDirect(bytes.length).order(ByteOrder.nativeOrder()).put(bytes);
            data.flip();

            return data;
        } catch (Exception e) {
            ShindoLogger.error("Failed to load resource", e);
        }

        return null;
    }

    public static ByteBuffer resourceToByteBuffer(File file) {

        try {
            byte[] bytes = org.apache.commons.io.IOUtils.toByteArray(Files.newInputStream(file.toPath()));

            ByteBuffer data = ByteBuffer.allocateDirect(bytes.length).order(ByteOrder.nativeOrder()).put(bytes);
            data.flip();

            return data;
        } catch (Exception e) {
            ShindoLogger.error("Failed to load resource", e);
        }

        return null;
    }
}

File: src/main/kotlin/me/miki/shindo/utils/IOUtils.kt
package me.miki.shindo.utils

import me.miki.shindo.logger.ShindoLogger
import me.miki.shindo.utils.transferable.ImageTransferable
import net.minecraft.client.Minecraft
import net.minecraft.util.ResourceLocation
import java.awt.Image
import java.awt.Toolkit
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.StringSelection
import java.io.File
import java.nio.Buffer
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files

object IOUtils {
    private val mc: Minecraft = Minecraft.getMinecraft()

    @JvmStatic
    fun copyStringToClipboard(s: String) {
        val stringSelection = StringSelection(s)
        toolkit.systemClipboard.setContents(stringSelection, null)
    }

    @JvmStatic
    fun getStringFromClipboard(): String? =
        try {
            toolkit.systemClipboard
                .getContents(null)
                .getTransferData(DataFlavor.stringFlavor)
                .toString()
        } catch (_: Exception) {
            null
        }

    @JvmStatic
    fun copyImageToClipboard(image: Image) {
        val imageSelection = ImageTransferable(image)
        toolkit.systemClipboard.setContents(imageSelection, null)
    }

    @JvmStatic
    fun getImageFromClipboard(): Image? =
        try {
            toolkit.systemClipboard.getContents(null).getTransferData(DataFlavor.imageFlavor) as Image
        } catch (_: Exception) {
            null
        }

    private val toolkit: Toolkit
        get() = Toolkit.getDefaultToolkit()

    @JvmStatic
    fun resourceToByteBuffer(location: ResourceLocation): ByteBuffer? =
        try {
            val bytes =
                org.apache.commons.io.IOUtils
                    .toByteArray(mc.resourceManager.getResource(location).inputStream)
            val data = ByteBuffer.allocateDirect(bytes.size).order(ByteOrder.nativeOrder()).put(bytes)
            data.flip() as Buffer
            data
        } catch (e: Exception) {
            ShindoLogger.error("Failed to load resource", e)
            null
        }

    @JvmStatic
    fun resourceToByteBuffer(file: File): ByteBuffer? =
        try {
            val bytes =
                org.apache.commons.io.IOUtils
                    .toByteArray(Files.newInputStream(file.toPath()))
            val data = ByteBuffer.allocateDirect(bytes.size).order(ByteOrder.nativeOrder()).put(bytes)
            data.flip() as Buffer
            data
        } catch (e: Exception) {
            ShindoLogger.error("Failed to load resource", e)
            null
        }
}
